io/LasVLR.cpp

```diff
@@ -57,6 +62,115 @@ ILeStream& operator>>(ILeStream& in, LasVLR& v)
 }
 
 
+std::istream& operator>>(std::istream& in, LasVLR& v)
+{
+    NL::json j;
+    in >> j;
+
+    // Make sure there isn't stuff in the input stream after the JSON
+    // object.
+    char c;
+    do
+    {
+        in >> c;
+    } while (std::isspace(c));
+    if (!in.eof())
+        throw pdal_error("Invalid characters following LAS VLR JSON object.");
+
+    // We forced an EOF above, so clear the error state.
+    in.clear();
+    if (!j.is_object())
+        throw pdal_error("LAS VLR must be specified as a JSON object.");
+
+    std::string description;
+    std::string b64data;
+    std::string userId;
+    double recordId(std::numeric_limits<double>::quiet_NaN());
+    for (auto& el : j.items())
+    {
+        if (el.key() == "description")
+        {
+            if (!el.value().is_string()) 
+                throw pdal_error("LAS VLR description must be specified "
+                    "as a string.");
+            description = el.value().get<std::string>();
+            if (description.size() > 32)
+                throw pdal_error("LAS VLR description must be 32 characters "
+                    "or less.");
+        }
+        else if (el.key() == "record_id")
+        {
+            if (!el.value().is_number())
+                throw pdal_error("LAS VLR record ID must be specified as "
+                    "a number.");
+            recordId = el.value().get<double>();
+            if (recordId < 0 ||
+                recordId > (std::numeric_limits<uint16_t>::max)() ||
+                recordId != (uint16_t)recordId)
+                throw pdal_error("LAS VLR record ID must be an non-negative "
+                    "integer less than 65536.");
+        }
+        else if (el.key() == "user_id")
+        {
+            if (!el.value().is_string()) 
+                throw pdal_error("LAS VLR user ID must be specified "
+                    "as a string.");
+            userId = el.value().get<std::string>();
+            if (userId.size() > 16)
+                throw pdal_error("LAS VLR user ID must be 16 characters "
+                    "or less.");
+        }
+        else if (el.key() == "data")
+        {
+            if (!el.value().is_string())
+                throw pdal_error("LAS VLR data must be specified as "
+                    "a base64-encoded string.");
+            b64data = el.value().get<std::string>();
+        }
+        else
+            throw pdal_error("Invalid key '" + el.key() + "' in VLR "
+                "specificiation.");
+    }
+    if (b64data.empty())
+        throw pdal_error("LAS VLR must contain 'data' member.");
+    if (userId.empty())
+        throw pdal_error("LAS VLR must contain 'user_id' member.");
+    if (std::isnan(recordId))
+        recordId = 1;
+
+    v.m_userId = userId;
+    v.m_recordId = recordId;
+    v.m_description = description;
+    v.m_data = Utils::base64_decode(b64data);
```

Comment:
Do we need to protect against too large of a size?

Reply:
The data?

Reply:
Sorry, yeah. The data's size needs to be under a VLR.

Reply:
It can be as large as any 64 bit integer value, right?  So, no, if that's what you mean.

Reply:
EVLRs can be as large as any 64 bit integer value. VLRs can only be as large as a 16 bit value.

Reply:
True.

We treat everything as an EVLR if it's over the VLR limit.  If we can't do that (pre 1.4), then we throw an error at the time we try to add the VLR to the list of VLRs.  But I think things are fine here.

Reply:
Ok. Anyone pushing 4gb of base64 through pipeline JSON deserves their fate anyway.

Reply:
https://github.com/PDAL/PDAL/blob/227855a35532b7e978245a6e1a1072aa00bfd36a/io/LasWriter.cpp#L573-L593